test/lazy/test_ddp.py

```diff
@@ -0,0 +1,144 @@
+# Owner(s): ["oncall: jit"]
+
+import copy
+import os
+import sys
+import torch
+import torch._lazy
+import torch._lazy.ts_backend
+import torch.distributed as dist
+import torch.nn as nn
+import torch.optim as optim
+
+from datetime import timedelta
+from torch._utils_internal import TEST_MASTER_ADDR as MASTER_ADDR
+from torch._utils_internal import TEST_MASTER_PORT as MASTER_PORT
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.testing._internal.common_utils import run_tests, TEST_WITH_DEV_DBG_ASAN, NO_MULTIPROCESSING_SPAWN
+from torch.testing._internal.common_distributed import MultiProcessTestCase
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if not torch.cuda.is_available() or not dist.is_nccl_available():
+    print("CUDA/NCCL not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print("Skip dev-asan as torch + multiprocessing spawn have known issues", file=sys.stderr)
+    sys.exit(0)
+
+if NO_MULTIPROCESSING_SPAWN:
+    print("Spawn not available, skipping tests.", file=sys.stderr)
+    sys.exit(0)
+
+torch._lazy.ts_backend.init()
+os.environ["WORLD_SIZE"] = str(torch.cuda.device_count())
+
+class ToyModel(nn.Module):
+    def __init__(self):
+        super(ToyModel, self).__init__()
+        self.net1 = nn.Linear(10, 10)
+        self.relu = nn.ReLU()
+        self.net2 = nn.Linear(10, 5)
+
+    def forward(self, x):
+        return self.net2(self.relu(self.net1(x)))
+
+
+def step(model, input, labels, device, loss_fn, optimizer):
+    input = input.to(device)
+    labels = labels.to(device)
+    optimizer.zero_grad()
+
+    output = model(input)
+
+    loss = loss_fn(output, labels)
+    loss.backward()
+    optimizer.step()
+    # FIXME(alanwaketan): Investigate why we produce different
+    # results if mark_step.
+    # if device.type == "lazy":
+    #     torch._lazy.mark_step(str(device))
```

Comment:
If I understand correctly the issue here is the `view` relationship will be break after `mark_step`. However if this is the step end I think that should be fine? We generally don't expect tensors' life time exceed the `mark_step`.

Reply:
it looks like the mark_step here is the only place you do a mark step.

if you do not do a mark step, then you rely on calling .cpu() when you pass loss to allclose, and then another .cpu() when you pass model parameters to allclose.

in theory calling mark step first should only make those two .cpu() more predictable, they should both return 'already cached' results.
* TODO: can you confirm that when mark_step is used, both .cpu() calls are actually grabbing cached data?

without calling mark_step, i guess each .cpu() call below would re-execute things.  It seems like that should still be correct.

I think you also said there are still graph-splits in this program.  It would probably be worth simplifying things until you can get it into one graph, just for simpler debugging.

Reply:
it looks like the mark_step here is the only place you do a mark step.

if you do not do a mark step, then you rely on calling .cpu() when you pass loss to allclose, and then another .cpu() when you pass model parameters to allclose.

in theory calling mark step first should only make those two .cpu() more predictable, they should both return 'already cached' results.
* TODO: can you confirm that when mark_step is used, both .cpu() calls are actually grabbing cached data?

without calling mark_step, i guess each .cpu() call below would re-execute things.  It seems like that should still be correct.

I think you also said there are still graph-splits in this program.  It would probably be worth simplifying things until you can get it into one graph, just for simpler debugging.